don't persist changed entities + autofill slots for policy entities

wochinge opened the pull request:
**Proposed changes**:
- don't persist `UserUttered` attribute which were changed by further events
- The tracker we return in the API / send to the action server now contains the combined `UserUttered.parse_data` in its field `latest_message`.`Combined` means that it contains the entities of NLU predictions and entities predicted by the policies. This allows users to access the entities as usual.
- support autofilling slots with entities predicted by policies
- rename `DefinePrevUserUtteredEntities` to `EntitiesAdded`

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

wochinge commented on tests/shared/core/test_trackers.py:
```diff
@@ -36,6 +36,8 @@
     LegacyForm,
     LegacyFormValidation,
     LoopInterrupted,
+    DefinePrevUserUtteredFeaturization,
```
@Ghostvv @tmbo Brought up if we can come up with better names / names which generalize better:

How about 
* `UserUtteredFeaturizated`
* `EntitiesAdded`

Ghostvv replied:
In the redesigned system, when/if we have them as separate events, I think it is good names. In the current system, I'd keep `Define` ones. But if you think it is better to rename now, please go for it

wochinge replied:
@joejuzl What do you think?

joejuzl replied:
If these are the only two which behave in this way I can see an argument for having them have a common prefix.

wochinge replied:
Tom also meant the point that all events are using the past tense to stress the fact that they mirror something in the past. So I think it makes sense to at least take that into account.

wochinge replied:
And events are so essential that they'll basically stay around forever one they were added.  Hence, more general names could help us to adapt in the future.

Ghostvv replied:
taking into account new development around e2e entities, we can safely rename them

wochinge replied:
Do you mean we should rename them because a lot of things around entities are gonna change or did you mean there is no need to rename as things are gonna stay the same? 🙈

Ghostvv replied:
I meant we can rename them because for entities we're not "defining" previous useruttered entities anymore, so the name `DefinePrevUserUtteredFeaturization` is not accurate. And since you pointed out that deprecating event names is a pain, we might go with the names we like from the start

wochinge replied:
Ok, so how about `EntitiesAdded`?

Ghostvv replied:
let's add `e2e` in there? for explicitness

wochinge replied:
In theory it doesn't really matter where entities are coming from, no?

wochinge replied:
Eg. if we split up events in something like 

* `UserInputEvent` -> pure unchanged user input
* `NLUIntentEvent`
* `NLUEntityEvent`
* `PolicyPrediction`

then we could actually use `EntityAdded` both for NLU and Policy prediction

Ghostvv replied:
sounds good

wochinge replied:
done

Ghostvv commented:
I'm not sure I can understand. So entities will be added every time in applied events? So that users will see "normal" `UserUttered` in custom actions?

Ghostvv commented:
if this is the case, I believe the `if` on line 728 in `events.py` should be removed

wochinge commented:
> I'm not sure I can understand. So entities will be added every time in applied events? So that users will see "normal" UserUttered in custom actions?

Entities are added via `applied_to` as you implemented it. However, any changes to `UserUttered.entities` won't affect the serialized version of the event. This gives you the updated `entities` for the featurization while preserving the original event.
A user would only see the original `UserUttered` in a custom action (without entities added by policies).

Ghostvv commented:
> A user would only see the original UserUttered in a custom action (without entities added by policies).}

then it's not very useful, this was the reason for updating `UserUttered` event, so that these entities are perceived by users exactly the same as extracted by TED

> This gives you the updated entities for the featurization while preserving the original event.

we actually don't use e2e entities as features

wochinge commented:
> we actually don't use e2e entities as features

So only entities / roles / groups from the NLU classification are considered as features?

> then it's not very useful, this was the reason for updating UserUttered event, so that these entities are perceived by users exactly the same as extracted by TED

This needs to be done in a different place then. We can't mutate the event. What specific locations are you thinking of when you mean they should be perceived the same as extracted by TED? It should work out of the box for everything except for custom actions. 

We could use `applied_events` instead combined with a `copy.deepcopy` but then we'd need to make sure we'd call that whenever exposing the tracker to the user. Currently we're e.g. not calling `applied_events` before sending them to the action server.

Ghostvv commented:
> So only entities / roles / groups from the NLU classification are considered as features?

yes

wochinge commented:
@Ghostvv @joejuzl Updated according to what we discussed. How you can test this: 
1. Use the `e2e` example bot
2. train it
3. Send the message `Hello`
4. `GET <host>:<port>:/conversations/<conversation ID>/tracker` will return a tracker where `latest_message` contains the entities from the `TEDPolicy` while the actual `events` are untouched.

wochinge commented:
@Ghostvv We should in my opinion also updated [this check](https://github.com/RasaHQ/rasa/blob/12672a7da8298eac897bebb17253be484d90ac39/rasa/shared/core/trackers.py#L610) to work for the new event as well, shouldn't we?

joejuzl commented:
@wochinge 

> GET <host>:<port>:/conversations/<conversation ID>/tracker will return a tracker where latest_message contains the entities from the TEDPolicy while the actual events are untouched.

So the `DefinePrevUserUttered...` events now only exist in the tracker within the ensemble?

wochinge commented:
> So the DefinePrevUserUttered... events now only exist in the tracker within the ensemble?

Nope, it's a regular event as the others as well. We just changed that it does not longer affect the persisted version of `UserUttered`. In addition we changed the `latest_message` property of the tracker to make it easier for the users to access the combined entities (e.g. in custom actions)

joejuzl reviewed (approved):
I've tested it an the entities show as expected.

wochinge commented:
@joejuzl @Ghostvv Added support for autofilling slots based on entities predicted by polices. Please(re-)review

Ghostvv reviewed (changes requested):
This solution seems to me very convoluted. Why do we update previous `UserUttered` with entities? Then we use a lot of different `.copy()` to mitigate that. Isn't it simpler to store `e2e_entities` separate on the tracker and add them to `latest_message` data directly?

wochinge commented:
>  Then we use a lot of different .copy() to mitigate that.

actually just one 😆 

> Isn't it simpler to store e2e_entities separate on the tracker and add them to latest_message data directly?

The nice thing about having it in `entities` is that story printing would work as usual. It allows to interact with the event as it'd be one thing.

Ghostvv commented on rasa/shared/core/trackers.py:
```diff
@@ -600,9 +623,15 @@ def update(self, event: Event, domain: Optional[Domain] = None) -> None:
-        if domain and isinstance(event, UserUttered):
-            # store all entities as slots
-            for e in domain.slots_for_entities(event.parse_data["entities"]):
+        if domain and isinstance(event, (UserUttered, DefinePrevUserUtteredEntities)):
+            entities = event.entities
+            if isinstance(event, UserUttered):
+                # Rather get entities from `parse_data` as
+                # `DefinePrevUserUtteredEntities` might have already affected the
+                # `UserUttered.entities` attribute
+                entities = event.parse_data["entities"]
```
could you please unpack this convoluted `if`, I think checks for `UserUttered` and `DefinePrevUserUtteredEntities` can be done in parallel

Ghostvv replied:
actually, wouldn't `event.parse_data["entities"] == event.entities` for `DefinePrevUserUtteredEntities`?

also, I think we have a constant for `"entities"`, no?

Ghostvv replied:
`# DefinePrevUserUtteredEntities might have already affected the UserUttered.entities attribute`? what does it mean? why behavior is not deterministic?

wochinge replied:
> could you please unpack this convoluted if, I think checks for UserUttered and DefinePrevUserUtteredEntities can be done in parallel

I made it more explicit. 

> actually, wouldn't event.parse_data["entities"] == event.entities for DefinePrevUserUtteredEntities?

`DefinePrevUserUtteredEntities` doesn't have a `parse_data`?

> also, I think we have a constant for "entities", no?
 
done

Ghostvv replied:
> DefinePrevUserUtteredEntities doesn't have a parse_data?

ok

what about `might` why it is not deterministic?

wochinge replied:
I'm trying to play it save. Pretty sure that this would happen for  the `InMemoryTrackerStore`

Ghostvv replied:
ok, could you please add comment, so it'll be easier to modify this piece in the future

wochinge replied:
As you already merged, should I directly push the comment to `e2e` or are we good?

Ghostvv replied:
sorry, forgot about this one. as it is just a comment could you please directly merge to `e2e`.

wochinge replied:
done

Ghostvv reviewed (approved):
ok, if you think it is better 🚀

what about renaming? also left one refactoring comment

Ghostvv commented on rasa/shared/core/trackers.py:
```diff
@@ -624,12 +625,13 @@ def update(self, event: Event, domain: Optional[Domain] = None) -> None:
-                entities = event.parse_data["entities"]
+                entities = event.parse_data[ENTITIES]
+            else:
+                entities = event.entities
```
what about:
```
if domain:
   if isinstance(event, UserUttered):
   if isinstance(event, DefinePrevUserUtteredEntities):
```
?

wochinge replied:
I think that's style dependent and should be optional to address.

Ghostvv replied:
I mean... this is colossal `if domain and isinstance(event, (UserUttered, DefinePrevUserUtteredEntities)):` and then the solitary `else` is confusing

Ghostvv replied:
it takes more than a second to understand this `if` for someone else

Ghostvv replied:
I don't think it is style dependent, we should make sure that the code is as simple as possible to grasp by looking at it

wochinge replied:
I still think it's clear. Happy to hear @joejuzl 's opinion on it.

Ghostvv commented on rasa/shared/core/events.py:
```diff
@@ -686,21 +688,22 @@ def __eq__(self, other) -> bool:
 
-class DefinePrevUserUtteredEntities(DefinePrevUserUttered):
-    """Event that is used to set entities on a previous user uttered event."""
+class EntitiesAdded(DefinePrevUserUttered):
```
I'd say the super class `DefinePrevUserUttered` should be removed, what do you think?

wochinge replied:
kept it but made it a more general approach to deduplicate this behavior for other events as well

Ghostvv commented on rasa/shared/core/events.py:
```diff
@@ -345,6 +345,18 @@ def __eq__(self, other: Any) -> bool:
         return True
 
 
+class SkipEventInMDStoryMixin(Event, ABC):
```
do we need that? let's just add `as_story_string` to all the methods

Ghostvv replied:
@joejuzl what do you think. this method seems to me redundant but also increases cognitive complexity of events

joejuzl replied:
I think it's fine. It means you can easily identify events which won't be in the story rather than having to look at the implementation of the method in each event.

Ghostvv replied:
to me it contradicts `Occam’s razor` principle, but after all it's your part of code

Ghostvv replied:
for me personally, it makes it harder to understand, I see the method - I know how it is represented, looking at superclass requires more cognitive function

joejuzl replied:
I see arguments for both ways - but in the interest of speed (and since it's already done like this) - I think we should keep it as is.
